### PR TITLE
CSI-4229 create snapshot 2.0

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/array_action/array_mediator_svc.py [hunk 1]
```diff
@@ -1292,3 +1309,58 @@ def _get_host_name_if_equal(self, nvme_host, fc_host, iscsi_host):
...
+        try:
+            return self.client.svctask.addsnapshot(name=name, volumes=source_volume_id, pool=pool)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if is_warning_message(ex.my_message):
+                logger.warning("exception encountered while creating snapshot '{}': {}".format(name,
+                                                                                               ex.my_message))
+            else:
+                logger.error("cannot create snapshot {0}, Reason is: {1}".format(name, ex))
+                if OBJ_ALREADY_EXIST in ex.my_message:
+                    raise array_errors.SnapshotAlreadyExists(name, self.endpoint)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1323]: does this mean we [don't](https://github.com/IBM/ibm-block-csi-driver/blob/v1.9.0/controller/controller_server/csi_controller_server.py#L382) [return](https://github.com/IBM/ibm-block-csi-driver/blob/v1.9.0/controller/array_action/array_mediator_svc.py#L468) the correct status code for flashcopy 1.0 snapshots that already exist?

File: controller/array_action/array_mediator_svc.py [hunk 2]
```diff
@@ -1292,3 +1309,58 @@ def _get_host_name_if_equal(self, nvme_host, fc_host, iscsi_host):
...
+                if NAME_NOT_EXIST_OR_MEET_RULES in ex.my_message:
+                    raise array_errors.PoolDoesNotExist(pool, self.endpoint)
+                if NOT_ENOUGH_EXTENTS_IN_POOL_CREATE in ex.my_message:
+                    raise array_errors.NotEnoughSpaceInPool(id_or_name=pool)
+                if NOT_CHILD_POOL in ex.my_message:
+                    raise array_errors.PoolDoesNotExist(pool, self.endpoint)
+                if any(msg_id in ex.my_message for msg_id in (NON_ASCII_CHARS, INVALID_NAME, TOO_MANY_CHARS)):
+                    raise array_errors.IllegalObjectName(ex.my_message)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1331]: are all of these relevant for flashcopy 2.0?
if so, could we wrap them in a `_create_object(create_func, **kwargs)` method?

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:1331]: or maybe a `@create_object` decorator

File: controller/array_action/array_mediator_svc.py [hunk 3]
```diff
@@ -1292,3 +1309,58 @@ def _get_host_name_if_equal(self, nvme_host, fc_host, iscsi_host):
...
+                raise array_errors.IllegalObjectName(ex.my_message)
+            else:
+                raise ex
+        return None
+
+    def _get_cli_snapshot_by_id(self, snapshot_id):
+        return self._lsvolumesnapshot(object_id=snapshot_id)
+
+    def _get_cli_snapshot_by_name(self, snapshot_name):
+        filter_value = 'snapshot_name={}'.format(snapshot_name)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1357]: is `filtervalue` really necessary here? isn't `object_id` enough?

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:1357]: :(

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:1322]: done

File: controller/array_action/array_mediator_svc.py (hunk 2 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:1330]: CSI-4261

File: controller/array_action/array_mediator_svc.py [hunk 4]
```diff
@@ -719,9 +734,14 @@ def create_snapshot(self, volume_id, snapshot_name, space_efficiency, pool):
         logger.info("creating snapshot '{0}' from volume '{1}'".format(snapshot_name, volume_id))
         source_volume_name = self._get_volume_name_by_wwn(volume_id)
         source_cli_volume = self._get_cli_volume_in_pool_site(source_volume_name, pool)
-        target_cli_volume = self._create_snapshot(snapshot_name, source_cli_volume, space_efficiency, pool)
+        if self._is_addsnapshot_supported():
+            target_cli_snapshot = self._add_snapshot(snapshot_name, source_cli_volume, pool)
+            snapshot = self._generate_snapshot_response_from_cli_snapshot(target_cli_snapshot, source_cli_volume)
+        else:
+            target_cli_volume = self._create_snapshot(snapshot_name, source_cli_volume, space_efficiency, pool)
+            snapshot = self._generate_snapshot_response_from_cli_volume(target_cli_volume, source_cli_volume.vdisk_UID)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:742]:
```suggestion
        if self._is_addsnapshot_supported():
            target_cli_snapshot = self._add_snapshot(snapshot_name, source_cli_volume, pool)
            return self._generate_snapshot_response_from_cli_snapshot(target_cli_snapshot, source_cli_volume)
        target_cli_volume = self._create_snapshot(snapshot_name, source_cli_volume, space_efficiency, pool)
        return self._generate_snapshot_response_from_cli_volume(target_cli_volume, source_cli_volume.vdisk_UID)
```

File: controller/array_action/array_mediator_svc.py [hunk 5]
```diff
@@ -257,12 +258,20 @@ def _generate_volume_response(self, cli_volume):
...
-    def _generate_snapshot_response(self, cli_snapshot, source_id):
+    def _generate_snapshot_response_from_cli_volume(self, cli_volume, source_id):
+        return self._generate_snapshot_response(cli_volume.capacity, cli_volume.vdisk_UID, cli_volume.id,
+                                                cli_volume.name, source_id)
+
+    def _generate_snapshot_response_from_cli_snapshot(self, cli_snapshot, source_cli_volume):
+        return self._generate_snapshot_response(source_cli_volume.capacity, "", cli_snapshot.snapshot_id,
+                                                cli_snapshot.snapshot_name, source_cli_volume.vdisk_UID)
+
+    def _generate_snapshot_response(self, capacity, vdisk_uid, internal_id, name, source_id):
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:269]: what same?

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:269]: order of arguments

[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: done

**Review by matancarmeli7 — APPROVED** [review]
LGTM

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 6]
```diff
@@ -591,29 +622,68 @@ def test_create_snapshot_as_stretched_success(self):
...
+        self._prepare_mocks_for_create_snapshot_addsnapshot()
+        snapshot = self.svc.create_snapshot("source_volume_id", "test_snapshot", space_efficiency=None, pool="pool1")
+
+        self.assertEqual(snapshot.capacity_bytes, 1024)
+        self.svc.client.svctask.addsnapshot.assert_called_once_with(name='test_snapshot', volumes='test_id',
+                                                                    pool='pool1')
+        self.svc.client.svcinfo.lsvolumesnapshot.assert_called_once_with(object_id=0)
+        self.assertEqual(snapshot.array_type, 'SVC')
+        self.assertEqual(snapshot.id, '')
+        self.assertEqual(snapshot.internal_id, 'snapshot_id')
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:668]: as per our [conversation](https://stackoverflow.com/a/57059952) - do we want to use it correctly in this PR?
or should we refactor throughout the project in a separate ticket?

Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:668]: lets do it after #501 is merged

Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:668]: good idea

Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/array_mediator_svc.py [hunk 7]
```diff
@@ -1305,3 +1322,56 @@ def _get_host_name_if_equal(self, nvme_host, fc_host, iscsi_host):
...
+        id_start, id_end = message.find('[') + 1, message.find(']')
+        raw_id = message[id_start:id_end]
+        return int(raw_id)
+
+    def _lsvolumesnapshot(self, **kwargs):
+        try:
+            return self.client.svcinfo.lsvolumesnapshot(**kwargs).as_single_element
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if OBJ_NOT_FOUND in ex.my_message or NAME_NOT_EXIST_OR_MEET_RULES in ex.my_message:
+                logger.info("snapshot not found")
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1357]:
```suggestion
                logger.info("snapshot not found for args: {}".format(kwargs))
```

**Review by oriyarde — APPROVED** [review]
LGTM GJ!

**Pull request merged**